From 279ca188b6e44da26fae4491366cc21d66cd6406 Mon Sep 17 00:00:00 2001
From: Dusty Mabe <dusty@dustymabe.com>
Date: Tue, 11 Nov 2025 21:37:17 -0500
Subject: [PATCH] stages/mkfs.xfs: support setting agcount

For disk images (and the filesystems on them) that are created and later
grown there can be a performance issue if the agcount for XFS
filesystems increases many times because of the growfs.

In order to help reduce the chances of performance issues happening we
want to be able to set the agcount on newly created filesystems to
2 (current default is 4) [1].

[1] https://github.com/coreos/fedora-coreos-tracker/issues/1993#issuecomment-3184392894
---
 stages/org.osbuild.mkfs.xfs           | 6 +++++-
 stages/org.osbuild.mkfs.xfs.meta.json | 4 ++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/stages/org.osbuild.mkfs.xfs b/stages/org.osbuild.mkfs.xfs
index ebb414a3..b4481655 100755
--- a/stages/org.osbuild.mkfs.xfs
+++ b/stages/org.osbuild.mkfs.xfs
@@ -10,10 +10,14 @@ def main(devices, options):
 
     uuid = options["uuid"]
     label = options.get("label")
+    agcount = options.get("agcount")
     opts = []
 
     if label:
-        opts = ["-L", label]
+        opts.extend(["-L", label])
+
+    if agcount:
+        opts.extend(["-d", f"agcount={agcount}"])
 
     subprocess.run(["mkfs.xfs", "-m", f"uuid={uuid}"] + opts + [device],
                    encoding='utf8', check=True)
diff --git a/stages/org.osbuild.mkfs.xfs.meta.json b/stages/org.osbuild.mkfs.xfs.meta.json
index d832b77e..25f1da18 100644
--- a/stages/org.osbuild.mkfs.xfs.meta.json
+++ b/stages/org.osbuild.mkfs.xfs.meta.json
@@ -33,6 +33,10 @@
           "description": "Label for the file system",
           "type": "string",
           "maxLength": 12
+        },
+        "agcount": {
+          "description": "The number of allocation groups for the file system",
+          "type": "integer"
         }
       }
     }
-- 
2.51.0

